File: renderers/android/src/main/kotlin/dev/varn/gui/VarnNineSliceView.kt
package dev.varn.gui

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.util.LruCache

/**
 * The frames every nine-slice draws from, decoded once per file however many boxes are painted with one.
 *
 * A screen built out of artwork draws the same frame for every button on it, and decoding the file each
 * time a commit sets the source again is the same work done forty times over.
 */
internal object VarnFrames {
    private const val HELD = 8 * 1024

    private val held = object : LruCache<String, Bitmap>(HELD) {
        override fun sizeOf(key: String, value: Bitmap) = value.byteCount / 1024
    }

    fun read(path: String): Bitmap? {
        held.get(path)?.let { return it }

        val drawn = BitmapFactory.decodeFile(path) ?: return null

        held.put(path, drawn)
        return drawn
    }
}

/**
 * A box painted with a picture cut into nine, which is what a frame drawn from artwork is.
 *
 * The four corners are drawn at their own size, each edge is stretched along one axis and the middle is
 * stretched both ways. The platform's own `NinePatchDrawable` reads the cuts out of marker pixels compiled
 * into the file rather than taking them as numbers, so the pieces are drawn here instead, which is exact
 * and is the same arithmetic the other two renderers do.
 */
class VarnNineSliceView(context: Context) : VarnBoxView(context) {
    // The pieces are drawn without smoothing, the way the other two draw them: a frame is artwork, and
    // a two-pixel rule drawn larger and smoothed comes out as a smear rather than a rule.
    private val paint = Paint()
    private val from = Rect()
    private val onto = Rect()

    private var picture: Bitmap? = null
    private var perState = mapOf<String, Bitmap>()
    private var cuts = intArrayOf(0, 0, 0, 0)
    private var drawnAt = 1f

    init {
        setWillNotDraw(false)
    }

    fun setSource(path: String?) {
        picture = path?.let { VarnFrames.read(it) }
        invalidate()
    }

    /** Takes the artwork a state is drawn with, which is what makes a frame a button that presses. */
    fun setSources(named: Map<String, String>) {
        perState = named.mapNotNull { (state, path) -> VarnFrames.read(path)?.let { state to it } }.toMap()
        invalidate()
    }

    /** Answers the artwork for how the view is now, which is the plain one where there is none of its own. */
    private fun drawing(): Bitmap? {
        if (!isEnabled) {
            perState["disabled"]?.let { return it }
        }

        if (isPressed) {
            perState["pressed"]?.let { return it }
        }

        if (isHovered) {
            perState["hovered"]?.let { return it }
        }

        if (isFocused) {
            perState["focused"]?.let { return it }
        }

        return picture
    }

    override fun drawableStateChanged() {
        super.drawableStateChanged()
        invalidate()
    }

    /** Takes where the picture is cut, in its own pixels, as top, right, bottom and left. */
    fun setSlice(top: Int, right: Int, bottom: Int, left: Int) {
        cuts = intArrayOf(top, right, bottom, left)
        invalidate()
    }

    /** Says how many points one pixel of the border comes out at, which sets how thick it reads. */
    fun setSliceScale(scale: Float) {
        drawnAt = scale
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        val drawn = drawing() ?: return
        val (top, right, bottom, left) = cuts

        if (drawn.width <= left + right || drawn.height <= top + bottom) {
            from.set(0, 0, drawn.width, drawn.height)
            onto.set(0, 0, width, height)
            canvas.drawBitmap(drawn, from, onto, paint)
            return
        }

        val outTop = (top * drawnAt).toInt()
        val outRight = (right * drawnAt).toInt()
        val outBottom = (bottom * drawnAt).toInt()
        val outLeft = (left * drawnAt).toInt()

        val columns = listOf(
            intArrayOf(0, left, 0, outLeft),
            intArrayOf(left, drawn.width - right, outLeft, width - outRight),
            intArrayOf(drawn.width - right, drawn.width, width - outRight, width),
        )

        val rows = listOf(
            intArrayOf(0, top, 0, outTop),
            intArrayOf(top, drawn.height - bottom, outTop, height - outBottom),
            intArrayOf(drawn.height - bottom, drawn.height, height - outBottom, height),
        )

        for (column in columns) {
            for (row in rows) {
                if (column[3] <= column[2] || row[3] <= row[2]) {
                    continue
                }

                from.set(column[0], row[0], column[1], row[1])
                onto.set(column[2], row[2], column[3], row[3])
                canvas.drawBitmap(drawn, from, onto, paint)
            }
        }
    }
}

File: renderers/ios/VarnNineSliceView.swift
import UIKit

/// A box painted with a picture cut into nine, which is what a frame drawn from artwork is.
///
/// The platform has this already: a layer told where the middle of its contents is keeps the four corners
/// at their own size, stretches each edge along one axis and stretches the middle both ways. Drawing the
/// pieces by hand would be the same picture a commit later, since the layer resizes with the view and
/// nothing here has to hear about it.
///
/// It is a control rather than a plain view because a frame is how a button is drawn from artwork, and
/// the platform reports the two edges of a press through a control and through nothing else. A plain
/// view is told a press happened once the finger has already lifted, so a frame that goes down when it
/// is held would go down as it was let go.
final class VarnNineSliceView: UIControl {
    private var picture: UIImage?
    private var perState: [String: UIImage] = [:]
    private var cuts = UIEdgeInsets.zero
    private var drawnAt: CGFloat = 1

    override init(frame: CGRect) {
        super.init(frame: frame)

        layer.contentsGravity = .resize
        // The pieces are drawn without smoothing, the way the other two draw them: a frame is artwork,
        // and a two-pixel rule drawn larger and smoothed comes out as a smear rather than a rule.
        layer.magnificationFilter = .nearest
        layer.minificationFilter = .nearest
    }

    required init?(coder: NSCoder) { fatalError("not used") }

    func setSource(_ image: UIImage?) {
        picture = image
        redraw()
    }

    /// Takes the artwork a state is drawn with, which is what makes a frame a button that presses.
    func setSources(_ pictures: [String: UIImage]) {
        perState = pictures
        redraw()
    }

    /// Answers the artwork for how the control is now, which is the plain one where there is none of its own.
    private var drawing: UIImage? {
        if !isEnabled, let held = perState["disabled"] {
            return held
        }

        if isHighlighted, let held = perState["pressed"] {
            return held
        }

        if isFocused, let held = perState["focused"] {
            return held
        }

        return picture
    }

    override var isHighlighted: Bool {
        didSet { redraw() }
    }

    override var isEnabled: Bool {
        didSet { redraw() }
    }

    override func didUpdateFocus(in context: UIFocusUpdateContext, with coordinator: UIFocusAnimationCoordinator) {
        super.didUpdateFocus(in: context, with: coordinator)
        redraw()
    }

    func setSlice(_ insets: UIEdgeInsets) {
        cuts = insets
        redraw()
    }

    /// Says how many points one pixel of the border comes out at, which is what sets how thick it reads.
    func setSliceScale(_ scale: CGFloat) {
        drawnAt = scale
        redraw()
    }

    /// A picture cut wider than itself has nothing left in the middle to stretch, so it draws as itself.
    private func redraw() {
        guard let drawn = drawing?.cgImage else {
            layer.contents = nil
            return
        }

        let width = CGFloat(drawn.width)
        let height = CGFloat(drawn.height)

        guard width > cuts.left + cuts.right, height > cuts.top + cuts.bottom else {
            layer.contents = drawn
            layer.contentsCenter = CGRect(x: 0, y: 0, width: 1, height: 1)
            layer.contentsScale = 1 / drawnAt
            return
        }

        layer.contents = drawn
        layer.contentsCenter = CGRect(
            x: cuts.left / width,
            y: cuts.top / height,
            width: (width - cuts.left - cuts.right) / width,
            height: (height - cuts.top - cuts.bottom) / height
        )

        // A layer draws its contents at one pixel per `contentsScale` points and keeps the pieces outside
        // the middle at that size, so this is what decides how thick the border comes out.
        layer.contentsScale = 1 / drawnAt
    }
}
